mlrun/launcher/local.py

```diff
@@ -23,17 +40,291 @@ def verify_base_image(runtime):
+    ):
+        run = self._enrich_run(
+            runtime=runtime,
+            runspec=runspec,
+            project_name=project,
+            name=name,
+            params=params,
+            inputs=inputs,
+            returns=returns,
+            artifact_path=artifact_path,
+            notifications=notifications,
+        )
+        # TODO client-server separation, combine all validation under one method
+        runtime._validate_output_path(run)
+        runtime._validate_run_params(run.spec.parameters)
+
+        if "V3IO_USERNAME" in os.environ and "v3io_user" not in run.metadata.labels:
+            run.metadata.labels["v3io_user"] = os.environ.get("V3IO_USERNAME")
+
+        logger.info(
+            "Storing function",
+            name=run.metadata.name,
+            uid=run.metadata.uid,
+            db=runtime.spec.rundb,
+        )
+        self.store_function(runtime, run, run.metadata)
+
+        from mlrun.run import MLClientCtx
+
+        execution = MLClientCtx.from_dict(
+            run.to_dict(),
+            self.db,
+            autocommit=False,
+            is_api=False,
+            store_run=False,
+        )
+
+        from mlrun.runtimes.generators import get_generator
+
+        # create task generator (for child runs) from spec
+        task_generator = get_generator(run.spec, execution)
+        if task_generator:
+            # verify valid task parameters
+            tasks = task_generator.generate(run)
+            for task in tasks:
+                runtime._validate_run_params(task.spec.parameters)
+
+        # post verifications, store execution in db and run pre run hooks
+        execution.store_run()
+        self._pre_run(runtime, run, execution)  # hook for runtime specific prep
+
+        last_err = None
+        # If the runtime is nested, it means the hyper-run will run within a single instance of the run.
+        # So while in the API, we consider the hyper-run as a single run, and then in the runtime itself when the
+        # runtime is now a local runtime and therefore `self._is_nested == False`, we run each task as a separate run by
+        # using the task generator
+        # TODO client-server separation might not need the not runtime._is_nested anymore as this executed local func
+        if task_generator and not runtime._is_nested:
+            # multiple runs (based on hyper params or params file)
+            runner = runtime._run_many
+            if hasattr(runtime, "_parallel_run_many") and task_generator.use_parallel():
+                runner = runtime._parallel_run_many
+            results = runner(task_generator, execution, run)
+            mlrun.runtimes.utils.results_to_iter(results, run, execution)
+            result = execution.to_dict()
+            result = runtime._update_run_state(result, task=run)
+
+        else:
+            # single run
+            try:
+                resp = runtime._run(run, execution)
```

Comment:
seems weird to me that you removed to watch here and we should also remove it from the server code.
is the watch never needed?

Reply:
I'll add comment, watch is not needed in local and server side.
Local side because you actually wait because it's the user process that gets executed, so either way you'll "watch" the run.
On server side you won't watch because it's async execution and the server shouldn't watch, but rather the remote client(launcher) executing the job.